### PR TITLE
[Graphbolt] Gpu cache feature store

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Graphbolt GPU feature is implemented. I think a generic FeatureStore that can accept any kind of Features is needed so that the user can cache some of the features, such as features but not labels.

### CODE REVIEW

File: src/runtime/cuda/gpu_cache.cu [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        *rv = GpuCacheRef32(
+            std::make_shared<GpuCache<unsigned int>>(num_items, num_feats));
+      else
+        *rv = GpuCacheRef64(std::make_shared<GpuCache<long long>>(  // NOLINT
+            num_items, num_feats));
+    });
+
+DGL_REGISTER_GLOBAL("cuda._CAPI_DGLGpuCacheQuery")
+    .set_body([](DGLArgs args, DGLRetValue *rv) {
+      IdArray keys = args[1];
```
Comment (nv-dlasalle) [inline src/runtime/cuda/gpu_cache.cu:129]: Do we need a check here that the keys are GPU accessible (GPU memory or pinned)? Or alternatively in the Query() function above?

File: src/runtime/cuda/gpu_cache.cu [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        ret.push_back(Value(MakeValue(std::get<2>(result))));
+      }
+
+      *rv = ret;
+    });
+
+DGL_REGISTER_GLOBAL("cuda._CAPI_DGLGpuCacheReplace")
+    .set_body([](DGLArgs args, DGLRetValue *rv) {
+      IdArray keys = args[1];
+      NDArray values = args[2];
```
Comment (nv-dlasalle) [inline src/runtime/cuda/gpu_cache.cu:154]: Do we need a check here that the keys and values are GPU accessible (GPU memory or pinned)? Or alternatively in the Replace() function above?

**Review by nv-dlasalle — APPROVED** [review]
Looks good to me. I would request though we add some checks so users will be notified by DGL that their input tensors are on the wrong device, rather than receiving an error from HugeCTR or CUDA.

File: src/runtime/cuda/gpu_cache.cu (hunk 2 above)
Comment (mfbalin) [reply src/runtime/cuda/gpu_cache.cu:154]: Sorry about leaving changes from #4341 here. Only the graphbolt python file is added in this PR. I will address these in #4341.

Comment (mfbalin) [reply src/runtime/cuda/gpu_cache.cu:154]: Committed the fixes in #4341.

**Comment by mfbalin** [comment]
@BarclayII Do you want to take a look as well?

File: python/dgl/graphbolt/impl/gpu_cached_feature_store.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+__all__ = ["GPUCachedFeatureStore"]
+
+
+class GPUCachedFeatureStore(FeatureStore):
+    r"""GPU cached feature store wrapping a fallback feature store."""
+
+    def __init__(self, fallback_store: FeatureStore, cache_size: int):
+        """Initialize GPU cached feature store with a given fallback.
+        Places the GPU cache to torch.cuda.current_device().
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: How can user set which device to load to? Will different GPUs share the same cache?

Comment (mfbalin) [reply python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: `torch.cuda.set_device(device)` before constructing the gpu cached feature store will ininitialize the GPU cache at the currently set device. Do we want the capability to pass the device, and if it is `None`, we fall back to the current_device?

Comment (mfbalin) [reply python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: I think `dgl.cuda.GPUCache` doesn't have an option to take in a device argument when constructing the GPU cache. It simply constructs it at the currently set CUDA device in the process, i.e. `CUDA_CALL(cudaGetDevice(&cuda_device));`

Comment (mfbalin) [reply python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: A single GPUCache instance belongs to a single GPU only. GPUs can't share the same cache. A wrapper for a multi-GPU cache could be implemented though.

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: I see, we can focus on single GPU cases for this PR, for multiple GPU use case, let's re-visit. Please put a comment on the header of the file to clarify it is designed for Single GPU usage.

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: Currently each GPU have a dataloader process, if torch.cuda.current_device() can recognize the device assigned to each process, then it should work properly.

Comment (mfbalin) [reply python/dgl/graphbolt/impl/gpu_cached_feature_store.py:16]: https://github.com/dmlc/dgl/blob/b6f5ba9a809fcd2e5b2c440f538c1ccc965a9e59/examples/multigpu/node_classification_sage.py#L242 sets the default device so it will just work out of the box.

**Comment by frozenbugs** [comment]
@mfbalin Are you still working on this PR?

**Comment by frozenbugs** [comment]
I see, we already have UVA work on our todolist, will kick of the work soon.

**Comment by mfbalin** [comment]
@frozenbugs I see that there is a feature store and a feature now. Feature store can hold multiple features. But there is no feature class to derive from. I need this functionality to add GPUCachedFeature. We also need to have a Generic FeatureStore implementation that can have different kinds of features inside of it so that some of the features can be the cached kind.

File: python/dgl/graphbolt/feature_store.py [hunk 4]
```diff
@@ -2,7 +2,46 @@
 
 import torch
 
-__all__ = ["FeatureStore"]
+__all__ = ["Feature", "FeatureStore"]
```
Comment (frozenbugs) [inline python/dgl/graphbolt/feature_store.py:5]: Can you do it in a separated PR?

Comment (frozenbugs) [reply python/dgl/graphbolt/feature_store.py:5]: I am doing it in https://github.com/dmlc/dgl/pull/6204, feel free to remove.

File: tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:1]: create a test_gpu_cached_feature_store.py and move the test there.

Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:1]: create a test_gpu_cached_feature.py and move the test there.

File: python/dgl/graphbolt/impl/gpu_cached_feature_store.py
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature_store.py:1]: rename to python/dgl/graphbolt/impl/gpu_cached_feature.py

File: python/dgl/graphbolt/impl/gpu_cached_feature_store.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        tensor([0, 1, 2])
+        >>> feature.update(torch.ones(3, dtype=torch.long),
+        ... torch.tensor([0, 1, 2]))
+        >>> feature.read(torch.tensor([0, 1, 2, 3]))
+        tensor([1, 1, 1, 3])
+        """
+        super(GPUCachedFeature, self).__init__()
+        assert isinstance(fallback_feature, Feature), (
+            f"fallback_feature must be FeatureStore, "
+            f"but got {type(fallback_feature)}."
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature_store.py:44]:
```suggestion
            f"The fallback_feature must be an instance of Feature, but got "
            f"{type(fallback_feature)}."
```

File: python/dgl/graphbolt/impl/gpu_cached_feature_store.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        )
+        self._fallback_feature = fallback_feature
+        # we query the underlying feature store to learn the feature dimension
+        self.cache_size = cache_size
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature_store.py:48]:
```suggestion
        self.cache_size = cache_size
        # Fetching the feature dimension from the underlying feature.
```

**Review by frozenbugs — APPROVED** [review]

File: tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    reason="GPUCachedFeature requires a GPU.",
+)
+def test_gpu_cached_feature():
+    a = torch.tensor([1, 2, 3]).to("cuda").float()
+    b = torch.tensor([[1, 2, 3], [4, 5, 6]]).to("cuda").float()
+
+    feat_store_a = gb.GPUCachedFeature(gb.TorchBasedFeature(a), 2)
+    feat_store_b = gb.GPUCachedFeature(gb.TorchBasedFeature(b), 1)
+
+    assert torch.equal(feat_store_a.read(), a)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py:21]:
```suggestion
    # Test read the entire feature.
    assert torch.equal(feat_store_a.read(), a)
```

File: tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    assert torch.equal(feat_store_b.read(), b)
+    assert torch.equal(
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py:23]:
```suggestion

    # Test read with ids.
    assert torch.equal(
```

File: tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    )
+    assert torch.equal(
+        feat_store_a.read(torch.tensor([1, 1]).to("cuda")),
+        torch.tensor([2.0, 2.0]).to("cuda"),
+    )
+    assert torch.equal(
+        feat_store_b.read(torch.tensor([1]).to("cuda")),
+        torch.tensor([[4.0, 5.0, 6.0]]).to("cuda"),
+    )
+    feat_store_a.update(
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py:35]:
```suggestion

    # Test update the entire feature.
    feat_store_a.update(
```

File: tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        torch.tensor([0.0, 1.0, 2.0]).to("cuda"),
+        torch.tensor([0, 1, 2]).to("cuda"),
+    )
+    assert torch.equal(
+        feat_store_a.read(), torch.tensor([0.0, 1.0, 2.0]).to("cuda")
+    )
+    feat_store_a.update(
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py:42]:
```suggestion

    # Test update with ids.
    feat_store_a.update(
```
